kuma/attachments/views.py

```diff
@@ -38,18 +40,33 @@ def raw_file(request, attachment_id, filename):
     """
     qs = Attachment.objects.select_related('current_revision')
     attachment = get_object_or_404(qs, pk=attachment_id)
-    if attachment.current_revision is None:
+    rev = attachment.current_revision
+    if rev is None:
         raise Http404
 
-    if is_untrusted(request):
-        rev = attachment.current_revision
-        response = StreamingHttpResponse(rev.file, content_type=rev.mime_type)
-        response['Content-Length'] = rev.file.size
+    # Attachments must be served from safe (untrusted) domains
+    if not is_untrusted(request):
+        return redirect(attachment.get_file_url(), permanent=True)
+
+    # NOTE: All of this, just to support conditional requests (last-modified / if-modified-since)
+    # Very important while we're potentially serving attachments from disk.
+    # Far less important when we're just redirecting to S3.
+    # Consider removing?
```

Comment:
I'd much rather simplify it and "outsource" as much as possible to S3. Your code is good but theirs is *really* good. 
So something like this:

```python
if settings.ATTACHMENTS_USE_S3:
    return redirect(rev.file.url)

if_modified_since = parse_http_date_safe(request.META.get('HTTP_IF_MODIFIED_SINCE'))
if if_modified_since and if_modified_since >= calendar.timegm(rev.created.utctimetuple()):
    response = HttpResponseNotModified()
else:
    response = StreamingHttpResponse(rev.file, content_type=rev.mime_type)
    response['Content-Length'] = rev.file.size
response['Last-Modified'] = convert_to_http_date(rev.created)
response['X-Frame-Options'] = f'ALLOW-FROM {settings.DOMAIN}'
return response
```

It's your call but I get a bit nervous about messing with the headers of a file that's going to be redirected anyway.

Reply:
This will need further revision once we're on S3, but here's my current thinking:

The attachment URLs we embed in documents are keyed to an `Attachment` ID, rather than a more specific `AttachmentRevision`.

That means that users need to make two requests for every attachment on a page:

1. Discover the current redirect
2. Fetch the actual attachment

The exchange ends up looking like this:

![redirect](https://user-images.githubusercontent.com/24193/70066484-b7e22780-15e4-11ea-969a-387e7b9f4984.png)

Because the CDN URL is unique for each `AttachmentRevision`, we can instruct browsers to cache them immutably. Users with a warm cache can then completely skip the second request.

The problem is the first request, used to discover the CDN URL. Is it worth caching that? How and for how long?

(I think it is: if we don't, then we have to handle resolving every attachment. And that could add significant latency for a user with an otherwise warm cache.)

Right now I believe it will just be cached for however long our default TTL is in CloudFront, but we should probably be more explicit about the caching headers once we're using S3 in production.

Reply:
I think my suggestion was to get out of the business of `last-modified-if` stuff if it is a redirect. 
The caching of the 302 will be cached by CloudFront too. 
```
curl -v https://developer.m.o/files/123/fw.png
< HTTP/2 302
< x-cache: Hit from cloudfront
< cache-control: max-age=900, public
...
```
(the 900 comes from https://github.com/mdn/kuma/blob/2d8ead0cff04e395366637aa7aaebf277d797e75/kuma/attachments/views.py#L36)
which means that even the redirect could be cached for 15 min. So for 99% of the time, in a given 15min period, our Django doesn't even need to bother with any of this stuff since CloudFront will take care of that too. 

I think we should just increase the number `15*60` to something much larger so that S3 caches the redirects for much longer. But that's rather immaterial. Having to figure out the redirect (in our Django) every 15min is peanuts compared to having to actually serve the file. 

So, perhaps leave your fancy if-modified-since stuff, as is, since it's rarely going to matter much anyway. And if so, perhaps remove the code comment about the note.